sktree/tree/_unsup_criterion.pyx

```diff
@@ -136,35 +164,165 @@ cdef class TwoMeans(UnsupervisedCriterion):
             The new ending position for which to move samples from the right
             child to the left child.
         """
-        # need to implement
-        pass
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+
+        cdef const SIZE_t[:] sample_indices = self.sample_indices
+        cdef const DOUBLE_t[:] sample_weight = self.sample_weight
+
+        cdef SIZE_t i
+        cdef SIZE_t p
+        cdef DOUBLE_t w = 1.0
+
+        # Update statistics up to new_pos
+        #
+        # Given that
+        #   sum_left[x] +  sum_right[x] = sum_total[x]
+        # and that sum_total is known, we are going to update
+        # sum_left from the direction that require the least amount
+        # of computations, i.e. from pos to new_pos or from end to new_po.
+        if (new_pos - pos) <= (end - new_pos):
+            for p in range(pos, new_pos):
+                i = sample_indices[p]
+
+                if sample_weight is not None:
+                    w = sample_weight[i]
+
+                self.sum_left += w * self.X[i]
```

Comment:
`X` is a 2D array, so I think this wouldn't work?

Reply:
Moreover, the code in scikit-learn for defining `sum_left, sum_right, sum_total`:

```
       # Count labels for each output
        self.sum_total = np.zeros((n_outputs, max_n_classes), dtype=np.float64)
        self.sum_left = np.zeros((n_outputs, max_n_classes), dtype=np.float64)
        self.sum_right = np.zeros((n_outputs, max_n_classes), dtype=np.float64)
```

So sum_left and sum_right have shape # outputs X # of classes. We don't have classes, or # of outputs. In the analogous code for `update()`, the code is:

```
                for k in range(self.n_outputs):
                    self.sum_left[k, <SIZE_t> self.y[i, k]] += w
```

So `sum_left` is just getting values of `w`. What's the idea for adding `w * X[i]`?